packages/env/lib/config.js

```diff
@@ -0,0 +1,197 @@
+			} else {
+				throw new ValidationError( `Could not read .wp-env.json: ${ error.message }` );
+			}
+		}
+
+		if ( config === null ) {
+			const type = await detectDirectoryType( configDirectoryPath );
+			if ( type === 'core' ) {
+				config = { core: '.' };
+			} else if ( type === 'plugin' ) {
+				config = { plugins: [ '.' ] };
+			} else if ( type === 'theme' ) {
+				config = { themes: [ '.' ] };
+			} else {
+				throw new ValidationError(
+					`No .wp-env.json file found at '${ configPath }' and could not determine if '${ configDirectoryPath }' is a WordPress installation, a plugin, or a theme.`
+				);
+			}
+		}
+
+		config = Object.assign(
+			{
+				core: null,
+				plugins: [],
+				themes: [],
+			},
+			config
+		);
+
+		if ( config.core !== null && typeof config.core !== 'string' ) {
+			throw new ValidationError( 'Invalid .wp-env.json: "core" must be null or a string.' );
+		}
+
+		if (
+			! Array.isArray( config.plugins ) ||
+			config.plugins.some( ( plugin ) => typeof plugin !== 'string' )
+		) {
+			throw new ValidationError( 'Invalid .wp-env.json: "plugins" must be an array of strings.' );
+		}
+
+		if (
+			! Array.isArray( config.themes ) ||
+			config.themes.some( ( theme ) => typeof theme !== 'string' )
+		) {
+			throw new ValidationError( 'Invalid .wp-env.json: "themes" must be an array of strings.' );
+		}
+
+		const workDirectoryPath = path.resolve( os.homedir(), '.wp-env', md5( configPath ) );
+
+		return {
+			name: path.basename( configDirectoryPath ),
+			configDirectoryPath,
+			workDirectoryPath,
+			dockerComposeConfigPath: path.resolve( workDirectoryPath, 'docker-compose.yml' ),
+			coreSource: parseSourceString( config.core, { hasTests: true, workDirectoryPath } ),
+			pluginSources: config.plugins.map( ( sourceString ) =>
+				parseSourceString( sourceString, { hasTests: false, workDirectoryPath } )
+			),
+			themeSources: config.themes.map( ( sourceString ) =>
+				parseSourceString( sourceString, { hasTests: false, workDirectoryPath } )
+			),
+		};
+	},
+};
+
+/**
+ * Parses a source string into a source object.
+ *
+ * @param {string} sourceString The source string. See README.md for documentation on valid source string patterns.
+ * @param {Object} options
+ * @param {boolean} options.hasTests Whether or not a `testsPath` is required. Only the 'core' source needs this.
+ * @param {string} options.workDirectoryPath Path to the work directory located in ~/.wp-env.
+ * @return {Source} A source object.
+ */
+function parseSourceString( sourceString, { hasTests, workDirectoryPath } ) {
+	if ( sourceString === null ) {
+		return null;
+	}
+
+	if (
+		sourceString.startsWith( '.' ) ||
+		sourceString.startsWith( HOME_PATH_PREFIX ) ||
+		path.isAbsolute( sourceString )
+	) {
+		let sourcePath;
+		if ( sourceString.startsWith( HOME_PATH_PREFIX ) ) {
+			sourcePath = path.resolve( os.homedir(), sourceString.substring( HOME_PATH_PREFIX.length ) );
```

Comment:
```suggestion
			sourcePath = path.resolve( os.homedir(), sourceString.slice( HOME_PATH_PREFIX.length ) );
```

`substring` has some strange nuance. I would make it a convention to use `slice` everywhere.

Reply:
Good suggestion! I'd love to see a lint rule about this.

Reply:
That would be great. @aduth What do you think?

Reply:
Seems reasonable to me, if even just for consistency's sake.

Not sure how you might implement it, at least not also without forbidding (albeit edge cases) of custom class functions of the same name

```js
class Foo {
    substring( a, b ) { /* ... */ }
}

const f = new Foo;
f.substring( 0, 0 ); // error?
```

Assuming some implementation like: https://astexplorer.net/#/gist/afd85b061baa01df071edab34fdc0647/cb262ba3b856120a3b4fe62411fbaa95edbc3caa

Reply:
>custom class functions of the same name

I don't think we have or should have those.